plugin-utilities-plugin/src/main/kotlin/org/jetbrains/research/pluginUtilities/preprocessing/Preprocessing.kt

```diff
@@ -0,0 +1,118 @@
+     * @param datasetDirectory A directory where each subdirectory is an individual repository
+     * @param outputDirectory A directory where preprocessed repositories will be stored.
+     * For each repository it creates a separate output subdirectory.
+     */
+    fun preprocessDataset(datasetDirectory: File, outputDirectory: File) {
+        for (repositoryRoot in datasetDirectory.subdirectories) {
+            val repositoryOutput = outputDirectory.resolve(repositoryRoot.name)
+            repositoryOutput.mkdir()
+            logger.info("Preprocessing repository ${repositoryRoot.name}")
+            preprocessRepository(repositoryRoot, repositoryOutput)
+        }
+    }
+}
+
+/**
+ * Object that preprocesses a given repository by mutating its file tree.
+ * For example, it can add new files to the repository, remove or change the existing ones.
+ * @property name The name of the preprocessing that is used for debugging and logging
+ */
+interface Preprocessing {
+    val name: String
+    fun preprocess(repoDirectory: File)
+}
+
+/**
+ * Adds local.properties files with sdk.dir=[androidSdkAbsolutePath] where it detects a Java build system
+ */
+class AndroidSdkPreprocessing(private val androidSdkAbsolutePath: String) : Preprocessing {
+    private val logger = Logger.getLogger(javaClass.name)
+
+    override val name: String = "Add Android SDK with local.properties"
+
+    companion object {
+        const val LOCAL_PROPERTIES_FILE_NAME = "local.properties"
+        const val SDK_PROPERTY_NAME = "sdk.dir"
+        val acceptedBuildSystems = listOf(BuildSystem.Gradle, BuildSystem.GradleKotlinDsl)
+    }
+
+    override fun preprocess(repoDirectory: File) {
+        repoDirectory.collectBuildSystemRoots(acceptedBuildSystems).forEach { projectRoot ->
+            logger.info("Updating $LOCAL_PROPERTIES_FILE_NAME in ${projectRoot.path}")
+            updateLocalProperties(projectRoot)
+        }
+    }
+
+    /**
+     * Sets the `sdk.dir` property in the local.properties file.
+     * If the file already exists, modifies it by adding the `sdk.dir` property or changing the existing value.
+     * Creates a new file with the new property otherwise.
+     */
+    private fun updateLocalProperties(projectRoot: File) {
+        val localPropertiesFile = projectRoot.resolve(LOCAL_PROPERTIES_FILE_NAME)
+        val properties = Properties()
+        if (localPropertiesFile.exists()) {
+            properties.load(localPropertiesFile.inputStream())
+        } else {
+            localPropertiesFile.createNewFile()
+        }
+        properties.setProperty(SDK_PROPERTY_NAME, androidSdkAbsolutePath)
+        properties.store(localPropertiesFile.outputStream(), null)
+    }
+}
+
+/**
+ * Removed directories from repositories with names [directoryNames]
+ * @param directoryNames The names of directories which are to be removed
+ */
+class DeleteDirectoriesPreprocessing(private val directoryNames: List<String>) : Preprocessing {
```

Comment:
Please, move these classes into several files - one file for android preprocessor, one for common and so on